flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java

```diff
@@ -177,65 +208,105 @@ protected boolean reconcileSpecChange(
 
         if (upgradeMode == UpgradeMode.STATELESS) {
             LOG.info("Stateless job, ready for upgrade");
-            return AvailableUpgradeMode.of(UpgradeMode.STATELESS);
+            return JobUpgrade.stateless(terminal);
         }
 
         var flinkService = ctx.getFlinkService();
-        if (ReconciliationUtils.isJobInTerminalState(status)
-                && !flinkService.isHaMetadataAvailable(ctx.getObserveConfig())) {
+        if (ReconciliationUtils.isJobCancelled(status)
+                || (terminal && !flinkService.isHaMetadataAvailable(ctx.getObserveConfig()))) {
 
             if (!SnapshotUtils.lastSavepointKnown(status)) {
-                throw new RecoveryFailureException(
+                throw new UpgradeFailureException(
                         "Job is in terminal state but last checkpoint is unknown, possibly due to an unrecoverable restore error. Manual restore required.",
                         "UpgradeFailed");
             }
-            LOG.info(
-                    "Job is in terminal state, ready for upgrade from observed latest checkpoint/savepoint");
-            return AvailableUpgradeMode.of(UpgradeMode.SAVEPOINT);
+            LOG.info("Job is in terminal state, ready for upgrade from observed latest state");
+            return JobUpgrade.savepoint(true);
         }
 
-        if (ReconciliationUtils.isJobRunning(status)) {
-            LOG.info("Job is in running state, ready for upgrade with {}", upgradeMode);
-            var changedToLastStateWithoutHa =
-                    ReconciliationUtils.isUpgradeModeChangedToLastStateAndHADisabledPreviously(
-                            resource, ctx.getObserveConfig());
-            if (changedToLastStateWithoutHa) {
-                LOG.info(
-                        "Using savepoint upgrade mode when switching to last-state without HA previously enabled");
-                return AvailableUpgradeMode.of(UpgradeMode.SAVEPOINT);
+        if (ReconciliationUtils.isJobCancelling(status)) {
+            LOG.info("Cancellation is in progress. Waiting for cancelled state.");
+            return JobUpgrade.pendingCancellation();
+        }
+
+        boolean running = ReconciliationUtils.isJobRunning(status);
+        boolean versionChanged =
+                flinkVersionChanged(
+                        ReconciliationUtils.getDeployedSpec(resource), resource.getSpec());
+
+        if (upgradeMode == UpgradeMode.SAVEPOINT) {
+            if (running) {
+                LOG.info("Job is in running state, ready for upgrade with savepoint");
+                return JobUpgrade.savepoint(false);
+            } else if (versionChanged
+                    || deployConfig.get(
+                            KubernetesOperatorConfigOptions
+                                    .OPERATOR_JOB_UPGRADE_LAST_STATE_FALLBACK_ENABLED)) {
+                LOG.info("Falling back to last-state upgrade mode from savepoint");
```

Comment:
I think this makes sense, considering that this condition would never change. A failing job may recover and savepoints would be possible but at a version change this is the only thing that is possible to do. 

To be honest it may make sense to completely remove the OPERATOR_JOB_UPGRADE_LAST_STATE_FALLBACK_ENABLED flag over time as falling back to the cancel based logic sounds reasonable in most cases

Reply:
Thank you for clarifying. +1 for removing the config option.